src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -151,57 +151,46 @@ void UpdateFieldStateOnWrite(IFieldSymbol field)
-                    field.ContainingType == owningSymbol.ContainingType)
-                {
-                    // For instance fields, ensure that the instance reference is being initialized by the constructor.
-                    var instanceFieldWrittenInCtor = isInConstructor &&
-                        fieldReference.Instance?.Kind == OperationKind.InstanceReference &&
-                        (!isFieldAssignemnt || fieldReference.Parent.Parent?.Kind != OperationKind.ObjectOrCollectionInitializer);
+                return false;
+            }
+
+            // Writes to fields inside constructor are ignored, except for the below cases:
+            //  1. Instance reference of an instance field being written is not the instance being initialized by the constructor.
+            //  2. Field is being written inside a lambda or local function.
+
+            // Check if we are in the constructor of the containing type of the written field.
+            var isInConstructor = owningSymbol.IsConstructor();
+            var isInStaticConstructor = owningSymbol.IsStaticConstructor();
+            var field = fieldReference.Field;
+            if ((isInConstructor || isInStaticConstructor) &&
+                field.ContainingType == owningSymbol.ContainingType)
+            {
+                // For instance fields, ensure that the instance reference is being initialized by the constructor.
+                var instanceFieldWrittenInCtor = isInConstructor &&
+                    fieldReference.Instance?.Kind == OperationKind.InstanceReference &&
+                    !fieldReference.IsTargetOfObjectMemberInitializer();
 
-                    // For static fields, ensure that we are in the static constructor.
-                    var staticFieldWrittenInStaticCtor = isInStaticConstructor && field.IsStatic;
+                // For static fields, ensure that we are in the static constructor.
+                var staticFieldWrittenInStaticCtor = isInStaticConstructor && field.IsStatic;
 
-                    if (instanceFieldWrittenInCtor || staticFieldWrittenInStaticCtor)
+                if (instanceFieldWrittenInCtor || staticFieldWrittenInStaticCtor)
+                {
+                    // Finally, ensure that the write is not inside a lambda or local function.
+                    if (!IsInAnonymousFunctionOrLocalFunction(fieldReference))
                     {
-                        // Finally, ensure that the write is not inside a lambda or local function.
-                        if (!IsInAnonymousFunctionOrLocalFunction(fieldReference))
-                        {
-                            // It is safe to ignore this write.
-                            return false;
-                        }
+                        // It is safe to ignore this write.
+                        return false;
                     }
                 }
-
-                return true;
             }
 
-            return false;
+            return true;
```

Comment:
is this fixing MakeFieldReadOnly?  If so, were there updated/added tests for tihs feature?

Reply:
No it is just a refactoring to use the newly added `GetValueUsageInfo(IOperation)` helper.

---
In reply to: [213410389](https://github.com/dotnet/roslyn/pull/29511#discussion_r213410389) [](ancestors = 213410389)